File: frontend/js/views/lease_option.js
// Lease-option (rent-to-own) generator — rent credits + net price at exercise,
// via /calc/lease-option. Previews live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderLeaseOption(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.lo.h1.title">// LEASE-OPTION (RENT-TO-OWN)</span></h1>
        <p class="muted small" data-i18n="view.lo.hint.intro">
            A tenant leases with an option to buy at an agreed price, with the up-front option fee and a
            portion of each rent payment credited toward the purchase. It computes the accumulated rent
            credits over the option period, the total credits, the net price at exercise, and the option
            expiration date. Drafting aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.lo.h2.inputs">Agreement details</h2>
            <form id="lo-form" class="inline-form">
                <label><span data-i18n="view.lo.label.state">State / jurisdiction</span>
                    <input type="text" name="state" value="Georgia" required></label>
                <label><span data-i18n="view.lo.label.landlord">Landlord / optionor</span>
                    <input type="text" name="landlord_name" value=""></label>
                <label><span data-i18n="view.lo.label.tenant">Tenant / optionee</span>
                    <input type="text" name="tenant_name" value=""></label>
                <label><span data-i18n="view.lo.label.property">Property address</span>
                    <input type="text" name="property_address" value=""></label>
                <label><span data-i18n="view.lo.label.fee">Option fee ($)</span>
                    <input type="number" step="100" min="0" name="option_fee_usd" value="5000" required></label>
                <label><span data-i18n="view.lo.label.credited">Option fee credited</span>
                    <input type="checkbox" name="option_fee_credited" checked></label>
                <label><span data-i18n="view.lo.label.rent">Monthly rent ($)</span>
                    <input type="number" step="0.01" min="0" name="monthly_rent_usd" value="2000" required></label>
                <label><span data-i18n="view.lo.label.credit">Monthly rent credit ($)</span>
                    <input type="number" step="0.01" min="0" name="monthly_rent_credit_usd" value="300"></label>
                <label><span data-i18n="view.lo.label.start">Option start date</span>
                    <input type="date" name="option_start_date" value="2026-08-01" required></label>
                <label><span data-i18n="view.lo.label.months">Option period (months)</span>
                    <input type="number" step="1" min="1" name="option_months" value="24" required></label>
                <label><span data-i18n="view.lo.label.price">Purchase price ($)</span>
                    <input type="number" step="1000" min="0" name="purchase_price_usd" value="350000" required></label>
                <label><span data-i18n="view.lo.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.lo.ph.statute'))}"></label>
            </form>
        </div>
        <div id="lo-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#lo-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            landlord_name: (fd.get('landlord_name') || '').trim(),
            tenant_name: (fd.get('tenant_name') || '').trim(),
            property_address: (fd.get('property_address') || '').trim(),
            option_fee_usd: Number(fd.get('option_fee_usd')) || 0,
            option_fee_credited: fd.get('option_fee_credited') != null,
            monthly_rent_usd: Number(fd.get('monthly_rent_usd')) || 0,
            monthly_rent_credit_usd: Number(fd.get('monthly_rent_credit_usd')) || 0,
            option_start_date: fd.get('option_start_date'),
            option_months: Math.round(Number(fd.get('option_months')) || 0),
            purchase_price_usd: Number(fd.get('purchase_price_usd')) || 0,
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcLeaseOption(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.lo.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#lo-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.lo.card.net">Net price at exercise</div>
                    <div class="value">${money(doc.net_price_at_exercise_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.lo.card.credits">Total credits</div>
                    <div class="value">${money(doc.total_credits_usd)}</div></div>
                <div class="card neg"><div class="label" data-i18n="view.lo.card.end">Option ends</div>
                    <div class="value">${esc(doc.option_end_date || '—')}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="lo-copy" type="button" data-i18n="view.lo.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="lo-download" type="button" data-i18n="view.lo.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#lo-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.lo.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.lo.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#lo-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'lease-option.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/lease_option.rs
//! Lease-option (rent-to-own) agreement — a tenant leases with an option to buy
//! at an agreed price, with the up-front option fee and a portion of each rent
//! payment credited toward the purchase. It computes the accumulated rent
//! credits over the option period, the total credits, the net price at exercise,
//! and the option expiration date. Distinct from a plain lease or purchase
//! agreement. Drafting aid, not legal advice.

use chrono::{Months, NaiveDate};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct LeaseOptionInput {
    pub landlord_name: String,
    pub tenant_name: String,
    pub property_address: String,
    pub option_fee_usd: f64,
    /// Whether the option fee is credited toward the purchase price.
    #[serde(default = "default_true")]
    pub option_fee_credited: bool,
    pub monthly_rent_usd: f64,
    /// Portion of each month's rent credited toward the purchase.
    #[serde(default)]
    pub monthly_rent_credit_usd: f64,
    pub option_start_date: String,
    pub option_months: u32,
    pub purchase_price_usd: f64,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

fn default_true() -> bool {
    true
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct LeaseOption {
    pub title: String,
    pub option_end_date: String,
    pub total_rent_credits_usd: f64,
    pub total_credits_usd: f64,
    pub purchase_price_usd: f64,
    pub net_price_at_exercise_usd: f64,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &LeaseOptionInput) -> LeaseOption {
    let total_rent_credits = cents(i.monthly_rent_credit_usd * i.option_months as f64);
    let fee_credit = if i.option_fee_credited { i.option_fee_usd } else { 0.0 };
    let total_credits = cents(total_rent_credits + fee_credit);
    let net_price = cents((i.purchase_price_usd - total_credits).max(0.0));

    let option_end = NaiveDate::parse_from_str(&i.option_start_date, "%Y-%m-%d")
        .ok()
        .and_then(|d| d.checked_add_months(Months::new(i.option_months)))
        .map(|d| d.format("%Y-%m-%d").to_string())
        .unwrap_or_default();

    let citation = i.statute_citation.trim();
    let pursuant = if citation.is_empty() {
        format!("This agreement is governed by the laws of the State of {}.", i.state)
    } else {
        format!("This agreement is governed by the laws of the State of {} ({}).", i.state, citation)
    };

    let fee_credit_note = if i.option_fee_credited {
        format!("The option fee of {} is credited toward the purchase price if the option is exercised.", money(i.option_fee_usd))
    } else {
        format!("The option fee of {} is non-refundable and is NOT credited toward the purchase price.", money(i.option_fee_usd))
    };

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Landlord/Optionor: {}\nTenant/Optionee: {}\nProperty: {}",
                i.landlord_name, i.tenant_name, i.property_address
            ),
        },
        DocClause {
            heading: "1. Grant of Option".into(),
            body: format!(
                "For an option fee of {}, the Landlord grants the Tenant the exclusive option to purchase the property, exercisable from {} through {} ({} months). {}",
                money(i.option_fee_usd), i.option_start_date, option_end, i.option_months, fee_credit_note
            ),
        },
        DocClause {
            heading: "2. Lease Terms".into(),
            body: format!(
                "During the option period the Tenant leases the property at {} per month, of which {} per month is credited toward the purchase price. Over {} months that is {} in rent credits.",
                money(i.monthly_rent_usd), money(i.monthly_rent_credit_usd), i.option_months, money(total_rent_credits)
            ),
        },
        DocClause {
            heading: "3. Purchase Price and Credits".into(),
            body: format!(
                "The agreed purchase price is {}. Total credits if the option is exercised at the end of the term: {} (rent credits {}{}). Net price at exercise: {}.",
                money(i.purchase_price_usd),
                money(total_credits),
                money(total_rent_credits),
                if i.option_fee_credited { format!(" + option fee {}", money(i.option_fee_usd)) } else { String::new() },
                money(net_price)
            ),
        },
        DocClause {
            heading: "4. Exercise".into(),
            body: "The Tenant may exercise the option by written notice to the Landlord before the expiration date and by proceeding to closing on customary terms. If the option is not exercised, it expires and the Landlord retains the option fee.".into(),
        },
        DocClause { heading: "5. Governing Law".into(), body: pursuant },
        DocClause {
            heading: "Signatures".into(),
            body: format!(
                "Landlord: ____________________  Date: __________\n{}\n\nTenant: ____________________  Date: __________\n{}",
                i.landlord_name, i.tenant_name
            ),
        },
    ];

    LeaseOption {
        title: "Lease-Option (Rent-to-Own) Agreement".into(),
        option_end_date: option_end,
        total_rent_credits_usd: total_rent_credits,
        total_credits_usd: total_credits,
        purchase_price_usd: i.purchase_price_usd,
        net_price_at_exercise_usd: net_price,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> LeaseOptionInput {
        LeaseOptionInput {
            landlord_name: "Owen Owner".into(),
            tenant_name: "Tina Tenant".into(),
            property_address: "7 Hill Rd".into(),
            option_fee_usd: 5_000.0,
            option_fee_credited: true,
            monthly_rent_usd: 2_000.0,
            monthly_rent_credit_usd: 300.0,
            option_start_date: "2026-08-01".into(),
            option_months: 24,
            purchase_price_usd: 350_000.0,
            state: "Georgia".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn credits_and_net_price() {
        let d = generate(&base());
        assert!(close(d.total_rent_credits_usd, 7_200.0));
        assert!(close(d.total_credits_usd, 12_200.0));
        assert!(close(d.net_price_at_exercise_usd, 337_800.0));
    }

    #[test]
    fn option_end_date() {
        // 2026-08-01 + 24 months = 2028-08-01.
        assert_eq!(generate(&base()).option_end_date, "2028-08-01");
    }

    #[test]
    fn uncredited_fee_excluded_from_credits() {
        let d = generate(&LeaseOptionInput { option_fee_credited: false, ..base() });
        // Only rent credits count: 7,200; net 342,800.
        assert!(close(d.total_credits_usd, 7_200.0));
        assert!(close(d.net_price_at_exercise_usd, 342_800.0));
        let c = d.clauses.iter().find(|c| c.heading == "1. Grant of Option").unwrap();
        assert!(c.body.contains("NOT credited"));
    }

    #[test]
    fn credit_clause_shows_breakdown() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading.contains("Purchase Price")).unwrap();
        assert!(c.body.contains("Net price at exercise: $337800.00"));
        assert!(c.body.contains("option fee $5000.00"));
    }

    #[test]
    fn lease_clause_shows_rent_credit() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading == "2. Lease Terms").unwrap();
        assert!(c.body.contains("$300.00 per month is credited"));
        assert!(c.body.contains("$7200.00 in rent credits"));
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&LeaseOptionInput { statute_citation: "O.C.G.A. § 44-7".into(), ..base() });
        assert_eq!(d.statutory_citation, "O.C.G.A. § 44-7");
        assert!(d.clauses.iter().any(|c| c.body.contains("O.C.G.A. § 44-7")));
    }

    #[test]
    fn bad_date_yields_empty_end() {
        let d = generate(&LeaseOptionInput { option_start_date: "x".into(), ..base() });
        assert_eq!(d.option_end_date, "");
    }
}
